module/state_synchronization/indexer/in_memory_indexer.go

```diff
@@ -0,0 +1,194 @@
+package indexer
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/convert"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data"
+	"github.com/onflow/flow-go/storage/store/inmemory/unsynchronized"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// InMemoryIndexer handles indexing of block execution data in memory.
+// It stores data in unsynchronized in-memory caches that are designed
+// to be populated once before being read.
+type InMemoryIndexer struct {
+	log             zerolog.Logger
+	metrics         module.ExecutionStateIndexerMetrics
+	registers       *unsynchronized.Registers
+	events          *unsynchronized.Events
+	collections     *unsynchronized.Collections
+	transactions    *unsynchronized.Transactions
+	results         *unsynchronized.LightTransactionResults
+	executionResult *flow.ExecutionResult
+	header          *flow.Header
+}
+
+// NewInMemoryIndexer creates a new indexer that uses in-memory storage implementations.
+// This is designed for processing unsealed blocks in the optimistic syncing pipeline.
+// The caches are created externally and passed to the indexer, as they will also be used
+// by the persister to save data permanently when a block is sealed.
+func NewInMemoryIndexer(
+	log zerolog.Logger,
+	metrics module.ExecutionStateIndexerMetrics,
+	registers *unsynchronized.Registers,
+	events *unsynchronized.Events,
+	collections *unsynchronized.Collections,
+	transactions *unsynchronized.Transactions,
+	results *unsynchronized.LightTransactionResults,
+	executionResult *flow.ExecutionResult,
+	header *flow.Header,
+) *InMemoryIndexer {
+	indexer := &InMemoryIndexer{
+		log:             log.With().Str("component", "in_memory_indexer").Logger(),
+		metrics:         metrics,
+		registers:       registers,
+		events:          events,
+		collections:     collections,
+		transactions:    transactions,
+		results:         results,
+		executionResult: executionResult,
+		header:          header,
+	}
+
+	indexer.metrics.InitializeLatestHeight(header.Height)
```

Comment:
we won't want to initialize the latest height in the indexer anymore since this is run more than once per runtime. let's remove this for now and we'll cleanup metrics in another PR